config/configmapprovider/composite.go

```diff
@@ -0,0 +1,280 @@
+// └─────────┬──────────┘
+//           │
+//           │   ┌─────────┬─────────┬─────────┬───┐
+//           └──►│Retrieved│Retrieved│Retrieved│...│
+//               └─────────┴─────────┴─────────┴───┘
+type compositeRetrieved struct {
+	// The current config map composed from subprovider's config maps. Will be returned
+	// by Get() func.
+	composedMap *config.Map
+
+	composeFunc configComposeFunc
+
+	// List of retrieved subproviders.
+	subRetrieved retrievedList
+
+	// Channel used to signal that one of the subprovider's Watchable's WaitForUpdate() returned.
+	readyWatchable chan readyWatchable
+
+	// A cancellable context that is used for long-running operations in WatchForUpdate().
+	watchContext       context.Context
+	cancelWatchContext func()
+}
+
+// A Watchable of which WaitForUpdate() returned and that is ready for Get() call.
+type readyWatchable struct {
+	// The Watchable that is ready.
+	watchable config.WatchableRetrieved
+
+	// The value that was returned by WaitForUpdate().
+	returnedError error
+}
+
+// Create a new composite Retrieved object from a list of Retrieved objects returned
+// by subproviders' Retrieve() calls.
+func newCompositeRetrieved(
+	ctx context.Context,
+	composeFunc configComposeFunc,
+	subRetrieved retrievedList,
+) (*compositeRetrieved, error) {
+	// Unfortunately config.Retrieved.Get() cannot return an error, but it is
+	// possible that composeFunc returns an error. Because of that we cannot
+	// do the composition using composeFunc in our Get() implementation and
+	// have to do it here so that we can return an error.
+	// TODO: change config.Retrieved.Get() to allow returning an error and move
+	// the following code to our Get().
+
+	// Call Get() on all sub-retrieved and combine the results using the composeFunc.
+	composedMap, err := subRetrieved.getAllAndCompose(composeFunc)
+	if err != nil {
+		// Didn't work. Close every WatchableRetrieved that may be on our list.
+		for _, retr := range subRetrieved {
+			if watchable, ok := retr.(config.WatchableRetrieved); ok {
+				err2 := watchable.Close(ctx)
+				err = multierr.Append(err, err2)
+			}
+		}
+		return nil, err
+	}
+
+	// Create a new cancellable context for long-running operations.
+	ctx, cancelFunc := context.WithCancel(context.Background())
+
+	m := &compositeRetrieved{
+		composedMap:        composedMap,
+		composeFunc:        composeFunc,
+		subRetrieved:       subRetrieved,
+		readyWatchable:     make(chan readyWatchable),
+		watchContext:       ctx,
+		cancelWatchContext: cancelFunc,
+	}
+
+	return m, nil
+}
+
+func (cr *compositeRetrieved) Get() *config.Map {
+	// WatchableRetrieved requires that Get() is never called concurrently
+	// with WatchForUpdate() which means a race on cr.composedMap is not possible
+	// since the only place where we modify it is in WatchForUpdate(), so there
+	// is no need to protect cr.composedMap from races.
+	return cr.composedMap
+}
+
+func (cr *compositeRetrieved) WatchForUpdate() error {
```

Comment:
I think a much simpler implementation also after I did https://github.com/open-telemetry/opentelemetry-collector/pull/4370 which solved few bugs, is to change the Provider interface and instead pass a "chan" or even better a func pointer:

```
// Provider is an interface that helps providing configuration's parser.
// Implementations may load the parser from a file, a database or any other source.
type Provider interface {
	// Retrieve goes to the configuration source and retrieves the selected data which
	// contains the value to be injected in the configuration and the corresponding watcher that
	// will be used to monitor for updates of the retrieved value.
	Retrieve(ctx context.Context, watcher func(Event{})) (Retrieved, error)

	// Close signals that the configuration for which it was used to retrieve values is no longer in use
	// and the object should close and release any watchers that it may have created.
	// This method must be called when the service ends, either in case of success or error.
	Shutdown(ctx context.Context) error
}

// Retrieved holds the result of a call to the Retrieve method of a Session object.
type Retrieved interface {
	// Get returns the Map.
	Get() *config.Map

	// Close signals that the configuration for which it was used to retrieve values is no longer in use
	// and the object should close and release any watchers that it may have created.
        // Implementations should not call the "watcher" func after this func returned.
	Close(ctx context.Context) error
}
```

Reply:
@bogdandrutu I posted a PR that does this, see https://github.com/open-telemetry/opentelemetry-collector/pull/4403
I like it, I believe it simplifies watching for changes. I suggest that we close this PR and continue further work with the callback approach.